runelite-client/src/main/java/net/runelite/client/plugins/keyremapping/KeyRemappingListener.java

```diff
@@ -0,0 +1,350 @@
+			{
+				case KeyEvent.VK_ENTER:
+				case KeyEvent.VK_SLASH:
+				case KeyEvent.VK_COLON:
+					// refocus chatbox
+					plugin.setTyping(true);
+					clientThread.invoke(() ->
+					{
+						plugin.unlockChat();
+					});
+					break;
+			}
+			
+		}
+		else
+		{
+			switch (e.getKeyCode())
+			{
+				case KeyEvent.VK_ENTER:
+					plugin.setTyping(false);
+					clientThread.invoke(() ->
+					{
+						plugin.lockChat();
+					});
+					break;
+				case KeyEvent.VK_ESCAPE:
+					plugin.setTyping(false);
+					clientThread.invoke(() ->
+					{
+						client.setVar(VarClientStr.CHATBOX_TYPED_TEXT, "");
+						plugin.lockChat();
+					});
+					break;
+				case KeyEvent.VK_BACK_SPACE:
+					if (Strings.isNullOrEmpty(client.getVar(VarClientStr.CHATBOX_TYPED_TEXT)))
+					{
+						plugin.setTyping(false);
+						clientThread.invoke(() -> plugin.lockChat());
+					}
+			}
+		}
+	}
+
+	@Override
+	public void keyReleased(KeyEvent e)
+	{
+		if (client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		if (plugin.chatboxFocused() && !plugin.isTyping())
+		{
+			modified.remove(e.getKeyCode());
+
+			if (config.wasdCameraRemap())
+			{
+				if (config.up().matches(e))
+				{
+					e.setKeyCode(KeyEvent.VK_UP);
+				}
+				else if (config.down().matches(e))
+				{
+					e.setKeyCode(KeyEvent.VK_DOWN);
+				}
+				else if (config.left().matches(e))
+				{
+					e.setKeyCode(KeyEvent.VK_LEFT);
+				}
+				else if (config.right().matches(e))
+				{
+					e.setKeyCode(KeyEvent.VK_RIGHT);
+				}
+				
+				
+				if (config.functionKeyRemap())
+				{
+					if (one().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F1);
+					}
+					else if (two().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F2);
+					}
+					else if (three().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F3);
+					}
+					else if (four().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F4);
+					}
+					else if (five().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F5);
+					}
+					else if (six().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F6);
+					}
+					else if (seven().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F7);
+					}
+					else if (eight().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F8);
+					}
+					else if (nine().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F9);
+					}
+					else if (ten().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F10);
+					}
+					else if (eleven().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F11);
+					}
+					else if (twelve().matches(e))
+					{
+						e.setKeyCode(KeyEvent.VK_F12);
+					}
+				}
+			}
+		}
+		else
+		{
+			// press d + enter + release d - causes the right arrow to never be released
+			Integer m = modified.get(e.getKeyCode());
+			if (m != null)
+			{
+				modified.remove(e.getKeyCode());
+				e.setKeyCode(m);
+			}
+		}
+	}
+	
+	private ModifierlessKeybind one()
+	{
+		return new ModifierlessKeybind(KeyEvent.VK_1, 0);
```

Comment:
Is there a reason you are making these each keystroke and not just keep keeping one of each? I think they are immutable.

Reply:
I modeled it after the wasdcamera config

Reply:
Wasd config is like that because it is config. This isn't config.